### PR TITLE
Fix33448

### LINKED ISSUE #33448: Regression in 3.6: Combination of union and intersection types with literal types not resolving specific types properly
**TypeScript Version:**  `3.6.2` and `3.6.3`

**Search Terms:** union types, intersection types, literal types

**Code**

```ts
type a = {
    type: 'a',
    data: string
}
type b = {
    type: 'b',
    name: string
}
type c = {
    type: 'c',
    other: string
}
type abc = a | b | c;
function f(problem: abc & (b | c)) {
    if (problem.type === 'b') {
        console.log(problem.name);
    } else {
        console.log(problem.other);
    }
}
```

**Expected behavior:**
TypeScript not to complain about the typings.

**Actual behavior:**
In the `f` function, TypeScript says that `problem` may not have `name` because type `Property 'name' does not exist on type 'a & c'`, but `a` and `c` are mutually exclusive types and therefore not possible. It also complains that `problem` may not have `other` because `Property 'other' does not exist on type 'a & b'`, but due to the if/else, it's not possible for `problem` to be of type `b`, and in fact the only possible type for it is `c`.

**Playground Link:** Since the playground is still on 3.5.1, this issue is not reproducible there. Here's a link to the example I posted above in the playground (though it is passing): https://www.typescriptlang.org/play/#code/C4TwDgpgBAhlC8UDeAoK6qkgLigchjwBo0MATGYGXAZ2ACcBLAOwHMUBfFLaAIwWSl0PXHl7EhUZjAC2EWgxbsuPKAGMBqDJnDz8aidoD2wABYR6Cpm07ddsXhsRwAPlH5u1AbhQAzAK7MasCMRsxQvgAUYPRGvAA2EDK4MI5QAGRQkR7qAJS5gtqMvlkxcYkyAHSq8LX44gVa2hhqYTRGiZXxRqzRsQlJldJyuT7aHFAQ8TTQTc3qbR0QXT195YMm5vSjklxcQA

**Related Issues:** https://github.com/microsoft/TypeScript/issues/33101 looked like it *might* be related, can't quite tell.

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #33448. Also fixes the issue in https://github.com/microsoft/TypeScript/pull/33498#issuecomment-533537286.

This PR supercedes #33498. (Thanks @jack-williams!)

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

**Comment by ahejlsberg** [comment]
RWC and Community test errors look to be preexisting conditions.

**Comment by weswigham** [comment]
RWC is clean on master. Maybe try syncing the branch and running again? Otherwise the change is real.

**Comment by ahejlsberg** [comment]
Looked at it more closely. The RWC difference is just simple error message change because we now pick a different (and arguably better) union constituent to elaborate.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -18310,19 +18310,14 @@ namespace ts {
...
-
         function isDiscriminantProperty(type: Type | undefined, name: __String) {
             if (type && type.flags & TypeFlags.Union) {
                 const prop = getUnionOrIntersectionProperty(<UnionType>type, name);
                 if (prop && getCheckFlags(prop) & CheckFlags.SyntheticProperty) {
                     if ((<TransientSymbol>prop).isDiscriminantProperty === undefined) {
                         (<TransientSymbol>prop).isDiscriminantProperty =
                             ((<TransientSymbol>prop).checkFlags & CheckFlags.Discriminant) === CheckFlags.Discriminant &&
-                            isDiscriminantType(getTypeOfSymbol(prop));
+                            !maybeTypeOfKind(getTypeOfSymbol(prop), TypeFlags.Instantiable);
```
Comment (weswigham) [inline src/compiler/checker.ts]: What's the justification for not maybe instantiables here? Previously it was "not generic indexes", so this is quite a bit broader an exclusion, if I'm not mistaken.

Comment (ahejlsberg) [reply src/compiler/checker.ts]: If you look at `isGenericIndexType` you'll see that it's more than just generic index types. The code change just clarifies what's actually going on, there's no change in behavior.

Comment (weswigham) [reply src/compiler/checker.ts]: Huh, alright, so it's equivalent. Neat.

**Pull request merged**